File: Rust/c-comparison/hello-world-c/hello_world.c
#include <linux/init.h>
#include <linux/module.h>
#include <linux/kernel.h>
   

MODULE_LICENSE("GPL");
MODULE_AUTHOR("Noé Terrier");	
MODULE_DESCRIPTION("C hello_world module");	


static int my_init(void)
{
    printk( KERN_NOTICE "Hello world from C module! (init)\n" );
    return  0;
}
   
static void my_exit(void)
{
    printk( KERN_NOTICE "Goodbye from C module! (exit)\n" );
    return;
}
   
module_init(my_init);
module_exit(my_exit);

File: Rust/c-comparison/hello-world-rust/hello_world.rs
// SPDX-License-Identifier: GPL-2.0

//! Rust out-of-tree sample

use kernel::prelude::*;

module! {
    type: HelloWorld,
    name: "hello_world",
    author: "Noe Terrier",
    description: "Rust hello_world module",
    license: "GPL",
}

struct HelloWorld;

impl kernel::Module for HelloWorld {
    fn init(_name: &'static CStr, _module: &'static ThisModule) -> Result<Self> {
        pr_info!("Hello world from Rust module! (init)\n");

        Ok(HelloWorld {})
    }
}

impl Drop for HelloWorld {
    fn drop(&mut self) {
        pr_info!("Goodbye from Rust module! (exit)\n");
    }
}
